From f73b161bd568a552f7e3ced2105eec37df642aac Mon Sep 17 00:00:00 2001
From: Tamizh Chelvam Raja <quic_tamizhr@quicinc.com>
Date: Thu, 19 Sep 2024 14:19:49 +0530
Subject: [PATCH] wifi: ath12k: Protect arsta tx stats access with spin lock

spin lock has been used while filling the arsta tx rate stats
and that should be used while fetching the value as well to
avoid reading incorrect value.

Signed-off-by: Tamizh Chelvam Raja <quic_tamizhr@quicinc.com>
Signed-off-by: Sarika Sharma <quic_sarishar@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -17633,11 +17633,14 @@ static void ath12k_mac_op_sta_statistics
 	sinfo->rx_duration = arsta->rx_duration;
 	sinfo->filled |= BIT_ULL(NL80211_STA_INFO_RX_DURATION);
 
+	spin_lock_bh(&ab->base_lock);
 	sinfo->tx_duration = arsta->tx_duration;
 	sinfo->filled |= BIT_ULL(NL80211_STA_INFO_TX_DURATION);
 
-	if (!arsta->txrate.legacy && !arsta->txrate.nss)
+	if (!arsta->txrate.legacy && !arsta->txrate.nss) {
+		spin_unlock_bh(&ab->base_lock);
 		return;
+	}
 
 	if (arsta->txrate.legacy) {
 		sinfo->txrate.legacy = arsta->txrate.legacy;
@@ -17651,6 +17654,7 @@ static void ath12k_mac_op_sta_statistics
 	}
 	sinfo->txrate.flags = arsta->txrate.flags;
 	sinfo->filled |= BIT_ULL(NL80211_STA_INFO_TX_BITRATE);
+	spin_unlock_bh(&ab->base_lock);
 
 	for (i = 0; i < ARRAY_SIZE(sinfo->chain_signal); i++) {
 		sinfo->chains &= ~BIT(i);
